### ISSUE TITLE
[App, API]: Random down-sampling makes it difficult to investigate short sections of a long time series

### ISSUE THREAD

**Issue by safelix** [wandb/wandb#5694]
### Current Behavior

Dear weights & biases team,

I have been using wandb extensively for a research project during the past year and was generally very happy with the entire setup. However, when it comes to viewing and collecting results, I believe there are still some unfortunate design choices, which make it quite difficult to investigate a short section of a long time series.

This concerns in particular the App. Using the zooming tool or adjusting the x-range manually, displays the new window as expected, but gives the impression of missing data points (see screenshots below). As discussed in #4177, this is due to the design choice of randomly down-sampling the data. I believe, however, that down-sampling should not lead to 'missing data points' in this case. On the contrary, the resolution should increase, because the number of samples should be fixed independently of the window size. This means in particular, that all data points should be resampled if the x-axis limits change. Otherwise, what is the point of zooming?

I know that one can retrieve data without sampling using `run.scan_history(min_step, max_step)` from the API, but this is of course less convenient and interactive than on the App. For custom, paper-ready plots, I find the API very useful and implemented my own subroutines. I personally prefer to down-sample in regular steps, but this requires iterating through the entire `SampledHistoryScan` and takes very long, in particular if a big sweep of runs is to be aggregated (one of my plots takes 2h to render). I don't know about the PROs/CONs of random vs regular subsampling, but from a user perspective I would highly appreciate the option to chose between the two. It would be very convenient to directly request slices of a timeseries using `[start:stop:step]` semantics or explicit indices in the API or even the App. Do you have any plans to support such a behavior in the future?

I'd like to tag @fdsig as you were already involved in #4177 and @vanpelt as this also concerns a fundamental design choice. I would be very happy to investigate this further with you.

Thanks a lot and best wishes, 
Felix

### Expected Behavior

Data points should not appear to be randomly missing. Either the density of data points should always be high enough and span the entire range. Or the missing information should be intuitive and interpretable: a low-resolution plot is easier to accept with equally spaced gaps.

### Steps To Reproduce

1. Open large timeseries
2. Change x-axis limits to small window
3. Observe 'missing data points'

### Screenshots

![image](https://github.com/wandb/wandb/assets/33009078/d030c2bc-f1a1-427d-b522-c60f7ebc34ea)
![image](https://github.com/wandb/wandb/assets/33009078/5a5ab2be-382e-4e8f-b4e8-1821f5f756c8)

### Environment

OS: Windows 

Browsers: Chrome

Version:

**Comment by nate-wandb** [comment]
Hi @safelix, thank you for taking the time to provide this feedback. So the main solution here would be to always return a set number of datapoints? I think this is an order of operations problem where currently we sample the steps THEN filter out steps that don't meet the x-axis requirements and the solution would be to first filter the points to the min/max x-axis and THEN sample if still necessary. 

> It would be very convenient to directly request slices of a timeseries using [start:stop:step] semantics or explicit indices in the API or even the App

`scan_history` does accept min_step= and max_step= args so are you saying that you would prefer list slicing here?

And I can make a feature request for the option to select normal sampling vs. random sampling. 

Let me know if I understand the requests here or if there is anything else I can answer. 

Thank you,
Nate

**Comment by safelix** [comment]
Hi @nate-wandb, thank you for reaching out. 

Yes, I believe your analysis and proposed solution are correct. If you first filter according to the x-axis min/max and then sample a fixed number of data points, I would expect to get the right result. The number of samples (let's call this `n_samples`) is probably already fixed internally, and it should also be fixed for every plot. I only see one edge case; if the filtered series contains less datapoints than `n_samples`, the entire series (of length <  `n_samples`) should be displayed. Further, it would be desirable to always display the first and last datapoint of the potentially filtered series and only randomly sample in between.

Sorry, I didn't mean to say that list slicing is necessarily the way to go. But having an argument for the `stepsize` (similar to slicing) in `scan_history` would speed up the runtime of the plots. 

Yes, I would appreciate if you made a feature request to select uniform vs random subsampling. For the API, this is equal to having the `stepsize` option in `scan_history` (potentially with slicing semantics). For the App, this is probably a bit more of an effort and I'm not sure whether this should be a local option of each plot, or a global option of an entire workspace (I'm leaning towards the second). And finally, if you implement this feature in the App, you could also consider to allow users to chose the `n_samples` manually (although this might have an effect on server loads).


Thanks a lot and best wishes,
Felix

**Comment by nate-wandb** [comment]
Hi @safelix, ok thank you for the feedback. So here are my main takeaways but feel free to add to this if I'm missing anything: 

1. Feature request for being able to choose normal/random sampling
2. Feature request to include first and last point in sampling **ALWAYS**
3. Feature request to first filter the x-min/x-max **BEFORE** deciding to sample further so this can be used to see a detailed subsection of the data.
4. Add stepsize arg (or support slicing implementation) to scan_history to allow users to customize how they sample instead of having to download the entire scan_history and then implementing this.  
5. Allow users to set `n_samples` in the app. (The current limit is 1500 and as you mentioned this may not be implemented because this is additional load on the database + the user's machine but I can still make this request)

Does this cover the main requests?

**Comment by safelix** [comment]
Yes, thank you! I think number 3 is most urgent, as this solves the zooming bug. Then I would prioritize in the order 2, 4, 1, 5.

**Comment by nate-wandb** [comment]
Ok, I can go ahead and submit these to the engineering team and follow up once they have a chance to look into this. Thank you for the detailed feedback!

**Comment by m-harmonic** [comment]
We are running into this issue when attempting to debug training issues that occur at certain unpredictable timesteps.  Having to use the API to retrieve the data makes it very difficult to use Wandb in the debugging process.

**Comment by cameronfr** [comment]
+1, trying to see exactly where the Infs start and cascade into a scaler death spiral (after 30k steps), and the sampling makes it hard and leading to wrong conclusions.
